File: sdks/js/tests/live/local.test.ts
import assert from "node:assert/strict";
import { after, before, describe, it } from "node:test";

import { Client, RouterError, type AgentHandle } from "../../src/index.js";
import { close, conversationModel, exhausted, uniqueId, unreachable } from "./target.js";

/**
 * The SDK against a router running on this machine.
 *
 * A local router takes a customer id at face value rather than a credential, because
 * nothing outside the machine can reach it, which is what makes this suite runnable
 * without secrets. What it proves is the part unit tests cannot: that the requests this
 * SDK builds are ones a real router accepts, and that what it says about conversations is
 * true of the backend rather than of the test server.
 */
const url = process.env["LOCAL_ACCELERATION_URL"] ?? "http://127.0.0.1:8098";
const customerId = process.env["LOCAL_ACCELERATION_CUSTOMER_ID"] ?? "support-local";

describe("the local router", { skip: await unreachable(url) }, () => {
  const api = new Client({ url, customerId });
  const opened: string[] = [];
  let model = "";
  /** An agent this router actually has configured, which is what a name resolves against. */
  let agent: AgentHandle | undefined;

  before(async () => {
    model = await conversationModel(api);
    const configs = await api.get("/v1/agents/configs").catch(() => []);
    const named = configs.find((config) => config.name);
    agent = named?.name ? api.agent(named.name) : undefined;
  });

  after(async () => {
    for (const id of opened) {
      await close(api, id);
    }
  });

  /** Remembers a session so a failing test does not leave one running. */
  function remember<T extends { id: string }>(session: T): T {
    if (!opened.includes(session.id)) {
      opened.push(session.id);
    }
    return session;
  }

  it("is healthy, and says what it depends on", async () => {
    const health = await api.get("/health");

    assert.equal(health.status, "ok");
    assert.equal(health.dependencies?.["llm"], "ok", "no conversation could be held");
  });

  it("resolves an agent config by the name it is stored under", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const config = await agent.config();

    assert.ok(config, `there is no config called ${agent.name}`);
    assert.equal(config.name, agent.name);
  });

  it("opens a conversation with labels, and reads them back", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const title = uniqueId("titled");
    const session = await agent.sessions.create({
      title,
      description: "opened by the local live suite",
      project: "docs",
      custom: { suite: "local" },
      persist_conversation: true,
      llm: model,
    });
    remember(session);

    assert.equal(session.created.title, title);
    assert.equal(session.created.project, "docs");
    assert.deepEqual(session.created.custom, { suite: "local" });
    assert.match(
      session.conversationId,
      /^agent:support-[0-9a-f-]{36}$/,
      "the backend names the channel, in the only shape it accepts",
    );

    await session.close();
  });

  it("finds a conversation again by what it was called", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const title = `Sendbird ${uniqueId("search")}`;
    const session = remember(
      await agent.sessions.create({ title, persist_conversation: true, llm: model }),
    );
    await session.close();

    const listed = await agent.sessions.query({ limit: 50 });
    assert.ok(
      listed.some((each) => each.id === session.id),
      "a conversation that ended is not being listed",
    );

    const found = await agent.sessions.search("Sendbird", { limit: 50 });
    assert.ok(
      found.some((each) => each.id === session.id),
      `searching for the title found nothing; ${found.length} other results`,
    );
  });

  it("keeps nothing about an incognito conversation", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const title = uniqueId("incognito");
    const session = remember(await agent.sessions.create({ incognito: true, title, llm: model }));
    assert.equal(session.created.incognito, true);
    assert.equal(session.conversationId, "", "an incognito session opens no channel");
    await session.close();

    const found = await agent.sessions.search(title, { limit: 50 });
    assert.equal(found.length, 0, "an incognito conversation was written down after all");
  });

  it("names the turn it answers, and writes down what the turn was made of", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const session = remember(
      await agent.sessions.create({ persist_conversation: true, llm: model }),
    );

    const answering = await session.responses.create("Reply with the single word: pong.");
    assert.ok(answering.id, "a recorded session names its turns");
    assert.equal(answering.status, "running");

    for await (const event of session.events()) {
      if (event.kind === "responded") {
        break;
      }
      if (event.kind === "error") {
        if (exhausted(event.error)) {
          t.skip(`nothing left to answer with: ${event.error}`);
          return;
        }
        assert.fail(`the session reported: ${event.error}`);
      }
    }

    const items = await answering.items.all();
    assert.ok(items.length > 0, "the turn was answered and nothing was written down");
    assert.equal(items[0]?.kind, "said", "every turn opens with what was asked");
    assert.ok(
      items.some((item) => item.kind === "answer"),
      `no answer was recorded; saw ${items.map((item) => item.kind).join(", ")}`,
    );

    await session.close();
  });

  it("forks a conversation into one of its own", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const parent = remember(
      await agent.sessions.create({
        title: "the first ask",
        project: "docs",
        persist_conversation: true,
        llm: model,
      }),
    );

    const forked = remember(await parent.fork({ title: "asked again" }));

    assert.notEqual(forked.id, parent.id);
    assert.equal(forked.created.forked_from, parent.id);
    assert.equal(forked.created.title, "asked again");
    assert.equal(forked.created.project, "docs", "what the fork did not mention it inherits");
    assert.notEqual(
      forked.conversationId,
      parent.conversationId,
      "a fork writes its own transcript",
    );

    await forked.close();
    await parent.close();
  });

  it("rewinds to a turn, taking the later ones out of the conversation", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const session = remember(await agent.sessions.create({ llm: model }));
    for (const question of ["Reply with the single word: one.", "Reply with the single word: two."]) {
      await session.responses.create(question);
      for await (const event of session.events()) {
        if (event.kind === "responded") {
          break;
        }
        if (event.kind === "error") {
          if (exhausted(event.error)) {
            t.skip(`nothing left to answer with: ${event.error}`);
            return;
          }
          assert.fail(`the session reported: ${event.error}`);
        }
      }
    }

    const [kept] = await session.responses.list();
    assert.ok(kept, "the first turn was not written down");
    await session.responses.rewind(kept);

    const left = await session.responses.list();
    assert.deepEqual(left.map((response) => response.id), [kept.id]);
    const forked = remember(await session.fork({ response_id: kept.id }));
    assert.equal(forked.created.forked_from, session.id);

    await forked.close();
    await session.close();
  });

  it("refuses to rewind a conversation kept in Stream Chat, which would bring the turns back", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const session = remember(await agent.sessions.create({ persist_conversation: true, llm: model }));

    await assert.rejects(
      () => session.responses.rewind("anything"),
      (raised: RouterError) => {
        assert.equal(raised.status, 400);
        assert.match(raised.message, /fork it at the response/);
        return true;
      },
    );

    await session.close();
  });

  it("refuses to fork an incognito conversation, since there is nothing to fork from", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const session = remember(await agent.sessions.create({ incognito: true, llm: model }));

    await assert.rejects(
      () => session.fork(),
      (raised: RouterError) => {
        assert.equal(raised.status, 400);
        assert.match(raised.message, /nothing to fork/);
        return true;
      },
    );

    await session.close();
  });

  it("stops listing a conversation as running once it is closed", async (t) => {
    if (!agent) {
      t.skip("this router has no agent configured to address by name");
      return;
    }

    const session = remember(await agent.sessions.create({ llm: model }));

    await session.close();

    const running = await agent.sessions.query({ state: "running", limit: 50 });
    assert.ok(
      !running.some((each) => each.id === session.id),
      "a closed session is still listed as running",
    );
  });
});

File: sdks/ruby/test/live/local_test.rb
# frozen_string_literal: true

require_relative "../test_helper"

# The SDK against a router that is actually running. Opt in with VISION_AGENTS_URL; the
# customer defaults to "examples". Run with `bundle exec rake live`.
#
# What it proves is the part the local router in the unit tests cannot: that the requests
# this gem builds are ones a real router accepts.
class TestLiveRouter < Minitest::Test
  URL = ENV.fetch("VISION_AGENTS_URL", "")
  CUSTOMER = ENV.fetch("VISION_AGENTS_CUSTOMER_ID", "examples")

  def setup
    skip "set VISION_AGENTS_URL to run against a router" if URL.empty?
    @api = VA::Client.new(url: URL, customer_id: CUSTOMER)
    @opened = []
  end

  def teardown
    @opened&.each(&:close)
  end

  def agent
    name = @api.get("/v1/agents/configs").map { |config| config["name"] }.compact.first
    skip "this router has no agent configured to address by name" unless name
    @api.agent(name)
  end

  # A conversation model this deployment resolves, since a name is resolved against the
  # deployment's own catalogue.
  def model
    @model ||= %w[llm-fast llm-thinking llm-flow].find do |target|
      @api.get("/v1/{modality}/routes/{target}", path: { modality: "llm", target: target })
    rescue VA::RouterError => e
      raise unless e.status == 404
    end || flunk("this deployment resolves none of the conversation models")
  end

  def open(**options)
    agent.sessions.create(llm: model, **options).tap { |session| @opened << session }
  end

  def answered(session)
    session.events.each do |event|
      return if event.kind == "responded"

      skip "nothing left to answer with: #{event.error}" if event.kind == "error" && event.error.include?("quota")
      flunk "the session reported: #{event.error}" if event.kind == "error"
    end
  end

  def unique(what)
    "ruby-#{what}-#{Time.now.to_i}-#{rand(1_000_000)}"
  end

  def test_the_router_is_healthy
    health = @api.get("/health")

    assert_equal "ok", health.dig("dependencies", "llm")
    assert_equal "ok", health["status"], "the router is #{health["status"]}: #{health["dependencies"]}"
  end

  def test_a_conversation_is_opened_with_labels_and_found_again
    title = "Sendbird #{unique("search")}"
    session = open(title: title, project: "docs", custom: { suite: "ruby" }, persist_conversation: true)
    session.close

    assert_equal title, session.created["title"]
    assert_equal({ "suite" => "ruby" }, session.created["custom"])
    assert_match(/\Aagent:/, session.conversation_id)
    assert(agent.sessions.query(limit: 50).any? { |each| each["id"] == session.id })
    assert(agent.sessions.search("Sendbird", limit: 50).any? { |each| each["id"] == session.id })
  end

  def test_an_incognito_conversation_keeps_nothing
    title = unique("incognito")
    session = open(incognito: true, title: title, persist_conversation: true)
    session.close

    assert_equal "", session.conversation_id
    assert_empty agent.sessions.search(title, limit: 50)
  end

  def test_a_turn_is_named_and_written_down
    session = open(persist_conversation: true)

    turn = session.responses.create("Reply with the single word: pong.")
    answered(session)

    refute_empty turn.id
    items = turn.items.all
    assert_equal "said", items.first["kind"]
    assert(items.any? { |item| item["kind"] == "answer" })
  end

  def test_rewinding_takes_the_later_turns_out_and_a_fork_starts_from_a_turn
    session = open
    ["Reply with the single word: one.", "Reply with the single word: two."].each do |question|
      session.responses.create(question)
      answered(session)
    end

    kept = session.responses.list.first
    session.responses.rewind(kept)
    assert_equal [kept["id"]], session.responses.list.map { |response| response["id"] }

    forked = session.fork(response_id: kept).tap { |fork| @opened << fork }
    assert_equal session.id, forked.created["forked_from"]
  end

  def test_a_conversation_kept_in_stream_chat_is_forked_rather_than_rewound
    session = open(persist_conversation: true)

    error = assert_raises(VA::RouterError) { session.responses.rewind("anything") }
    assert_equal 400, error.status

    forked = session.fork(title: "asked again").tap { |fork| @opened << fork }
    assert_equal session.id, forked.created["forked_from"]
    refute_equal session.conversation_id, forked.conversation_id
  end

  def test_a_closed_conversation_is_no_longer_running
    session = open
    session.close

    refute(agent.sessions.query(state: "running", limit: 50).any? { |each| each["id"] == session.id })
  end

  def test_a_guest_is_minted
    guest = @api.guest_user(name: "Ruby guest")

    refute_empty guest["id"]
    refute_empty guest["token"]
  end

  def test_a_folder_is_synced_once_and_then_read_back
    root = File.join(Dir.mktmpdir, unique("agent"))
    FileUtils.mkdir_p(root)
    File.write(File.join(root, "agent.yaml"), "llm: #{model}\n")
    File.write(File.join(root, "instructions.md"), "You are a test agent. Answer in one word.\n")

    first = VA::Agent.new(folder: root, client: @api).sync
    second = VA::Agent.new(folder: root, client: @api).sync

    assert_equal File.basename(root), first.dig("config", "name")
    assert_equal true, second["unchanged"]
    assert_equal first.dig("config", "id"), second.dig("config", "id")
  ensure
    id = first&.dig("config", "id")
    @api.delete("/v1/agents/configs/{id}", path: { id: id }) if id
    FileUtils.rm_rf(File.dirname(root)) if root
  end

  def test_a_worker_is_told_its_id
    dispatch = VA::Dispatch.new(client: @api).wait_for_call { nil }
    runner = Thread.new { dispatch.run }

    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + 10
    sleep 0.05 until dispatch.worker_id || Process.clock_gettime(Process::CLOCK_MONOTONIC) > deadline
    dispatch.stop
    runner.join(5)

    refute_nil dispatch.worker_id
  end

  def test_a_question_is_answered_over_the_llm_stream
    complete = VA::Router.new(client: @api).llm.realtime(target: model) do |llm|
      llm.respond("Reply with the single word: pong.")
    end

    refute_empty complete["text"]
  end

  def test_search_answers
    answer = VA::Router.new(client: @api).search("Stream Vision Agents", results: 3)

    refute_empty answer["results"]
  end
end
